### DISCUSSION TITLE
the problem in another machine

### DISCUSSION THREAD

**Discussion by Soikie** [pyinstaller/pyinstaller#7266] [Help]
I use pyinstaller for follow code, the exe works fine on the computer I use for programminge . But it raise this erroe when i run it on another computer ,the two computer is the same configuration.Of course,the yolov5.dll is actually in the path . At first ,i didnt add the dll to the binary in spec,i just copy the dll to the new  folder generated. After the error , i add the dll to the spec ,but i didnt find the dll in the folder generated. How can i make the pyinstaller find the dll?

error
`Traceback (most recent call last):
  File "PyInstaller\loader\pyimod03_ctypes.py", line 53, in __init__
  File "ctypes\__init__.py", line 382, in __init__
FileNotFoundError: Could not find module 'D:\python_trt\yolov5.dll' (or one of its dependencies). Try using the full path with constructor syntax.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "python_trt.py", line 34, in <module>
  File "python_trt.py", line 8, in __init__
  File "PyInstaller\loader\pyimod03_ctypes.py", line 55, in __init__
pyimod03_ctypes.install.<locals>.PyInstallerImportError: Failed to load dynlib/dll './yolov5.dll'. Most likely this dynlib/dll was not found when the application was frozen.
[10388] Failed to execute script 'python_trt' due to unhandled exception!`

code
`from ctypes import *
import cv2
import numpy as np
import numpy.ctypeslib as npct

class Detector():
    def __init__(self,model_path,dll_path):
        self.yolov5 = CDLL(dll_path)
        self.yolov5.Detect.argtypes = [c_void_p,c_int,c_int,POINTER(c_ubyte),npct.ndpointer(dtype = np.float32, ndim = 2, shape = (50, 6), flags="C_CONTIGUOUS")]
        self.yolov5.Init.restype = c_void_p
        self.yolov5.Init.argtypes = [c_void_p]
        self.yolov5.cuda_free.argtypes = [c_void_p]
        self.c_point = self.yolov5.Init(model_path)

    def predict(self,img):
        rows, cols = img.shape[0], img.shape[1]
        res_arr = np.zeros((50,6),dtype=np.float32)
        self.yolov5.Detect(self.c_point,c_int(rows), c_int(cols), img.ctypes.data_as(POINTER(c_ubyte)),res_arr)
        self.bbox_array = res_arr[~(res_arr==0).all(1)]
        return self.bbox_array

    def free(self):
        self.yolov5.cuda_free(self.c_point)

def visualize(img,bbox_array):
    for temp in bbox_array:
        bbox = [temp[0],temp[1],temp[2],temp[3]]  #xywh
        clas = int(temp[4])
        score = temp[5]
        cv2.rectangle(img,(int(temp[0]),int(temp[1])),(int(temp[0]+temp[2]),int(temp[1]+temp[3])), (105, 237, 249), 2)
        img = cv2.putText(img, "class:"+str(clas)+" "+str(round(score,2)), (int(temp[0]),int(temp[1])-5), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (105, 237, 249), 1)
    return img

det = Detector(model_path=b"./yolov5s.engine",dll_path="./yolov5.dll")  # b'' is needed
img = cv2.imread("./images/bus.jpg")
result = det.predict(img)
img = visualize(img,result)
cv2.imshow("img",img)
cv2.waitKey(0)
det.free()
cv2.destroyAllWindows()`

**Comment by rokm** [comment]
Is `yolov5.dll` collected in your application? Since it is loaded programmatically, PyInstaller will likely not collect for you automatically. So you either need to copy it manually to your application's top-level directory, or use [`--add-binary` command line switch](https://pyinstaller.org/en/stable/usage.html#cmdoption-add-binary). If you copy it manually, you will also need to ensure its dependencies are available.

> det = Detector(model_path=b"./yolov5s.engine",dll_path="./yolov5.dll")  # b'' is needed
> img = cv2.imread("./images/bus.jpg")

In general, you should avoid relative paths in your application, because this implicitly assumes that `yolov5s.engine`, `yolov5.dll`, and `images` directory are located in your current working directory. Which may or may not be the case. Instead, anchor them to the `__file__`, as described here: https://pyinstaller.org/en/stable/runtime-information.html#using-file

**Reply by Soikie** [reply]
sorry,I slipped up when I wrote this down,the yolov5.dll is actually in the path . At first ,i didnt add the dll to the binary in spec,i just copy the dll to the new folder generated. After the error , i add the dll to the spec ,but i didnt find the dll in the folder generated. How can i make the pyinstaller find the dll?

**Reply by rokm** [reply]
How did you add the dll to the spec? Can we see the spec?

**Reply by Soikie** [reply]
`# -*- mode: python ; coding: utf-8 -*-


block_cipher = None


a = Analysis(['python_trt.py'],
             pathex=['D:/PY/yolo_exe'],
             binaries=[('/yolov5.dll','.')],
             datas=[],
             hiddenimports=[],
             hookspath=[],
             hooksconfig={},
             runtime_hooks=[],
             excludes=[],
             win_no_prefer_redirects=False,
             win_private_assemblies=False,
             cipher=block_cipher,
             noarchive=False)
pyz = PYZ(a.pure, a.zipped_data,
             cipher=block_cipher)

exe = EXE(pyz,
          a.scripts, 
          [],
          exclude_binaries=True,
          name='python_trt',
          debug=False,
          bootloader_ignore_signals=False,
          strip=False,
          upx=True,
          console=True,
          disable_windowed_traceback=False,
          target_arch=None,
          codesign_identity=None,
          entitlements_file=None )
coll = COLLECT(exe,
               a.binaries,
               a.zipfiles,
               a.datas, 
               strip=False,
               upx=True,
               upx_exclude=[],
               name='python_trt')
`

**Reply by Soikie** [reply]
the structure is :
--yolo_exe
----python_trt.py
----yolov5s.engine
----yolov5.dll
----images

**Reply by rokm** [reply]
>  binaries=[('/yolov5.dll','.')],

You should provide the full path to `yolov5.dll`. So 
```
 binaries=[('D:/PY/yolo_exe/yolov5.dll','.')],
```

**Reply by Soikie** [reply]
Unfortunately, it still doesn't work.  I'm trying to add yolov5s.dll 's dependency to the folder .Then occur a new error. This seems to close to success

error
`PS D:\python_trt> D:\python_trt\python_trt.exe
Traceback (most recent call last):
  File "PyInstaller\loader\pyimod03_ctypes.py", line 53, in __init__
  File "ctypes\__init__.py", line 382, in __init__
OSError: [WinError 193] %1 is not a valid Win32 application。

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "python_trt.py", line 34, in <module>
  File "python_trt.py", line 8, in __init__
  File "PyInstaller\loader\pyimod03_ctypes.py", line 55, in __init__
pyimod03_ctypes.install.<locals>.PyInstallerImportError: Failed to load dynlib/dll './yolov5.dll'. Most likely this dynlib/dll was not found when the application was frozen.
[9688] Failed to execute script 'python_trt' due to unhandled exception!
PS D:\python_trt>`

**Reply by rokm** [reply]
> Failed to load dynlib/dll './yolov5.dll'. 

You should probably also replace relative those paths with absolute ones...

I.e., change
```
det = Detector(model_path=b"./yolov5s.engine",dll_path="./yolov5.dll") # b'' is needed
```

into

```python
app_root = os.path.dirname(__file__)
model_path = os.path.join(app_root, "yolov5s.engine").encode("utf8")
dll_path = os.path.join(app_root, "yolov5.dll")
det = Detector(model_path=model_path, dll_path=dll_path)
```

You'll also need to figure out what to do with images directory and use absolute path to it.

**Reply by rokm** [reply]
Also, are you mixing up 32-bit and 64-bit DLLs by any chance?

**Reply by Soikie** [reply]
I sadly find the yolov5.dll 's dependencies also have thier dependencies and these second  dependencies have their dependencies. I think there is only one way that is let pyinstaller to add these dlls. I will try change these paths

**Reply by Soikie** [reply]
Probably i need do some extra work if i use dll in my code ?

**Reply by rokm** [reply]
Depends on DLL and its dependencies, really. If they are link-time dependencies and are available in `PATH` during PyInstaller build, then adding it via `binaries` should also collect the dependencies.
